### ISSUE TITLE
Android support - Jaeger side

### ISSUE THREAD

**Issue by jpkrohling** [jaegertracing/jaeger#993]
Prerequisite: https://github.com/opentracing/opentracing-java/issues/296

Once the linked task is done, it would be a good idea to see what's missing on Jaeger's side: plug the Jaeger Client Java into the built app and get spans submitted to a remote collector (removing the agent from the equation). If it just works (I doubt it would), the next step would be to continue working on the OpenTracing side of the task, with the instrumentation libraries.

**Comment by jpkrohling** [comment]
That's wonderful! I outlined a few steps here: https://github.com/opentracing/opentracing-java/issues/300

This is mainly to assess that the base OT libraries do work out-of-the-box with Android applications. Once that is done, it would just be a matter of changing the actual tracer, from `MockTracer` to Jaeger's.

I believe the Jaeger Java Client will *not* work, but it would be really helpful if you could try it out and report what seems to be broken.

**Comment by Beanow** [comment]
Great that looks like things I can manage. I'll most likely start out instrumenting a button and an API call. See if I can get Traefik to pick this up as parent spans.

**Comment by jpkrohling** [comment]
I would keep Traefik out of the picture for now, as it's one more thing that can fail. With the MockTracer, you can check whether (and how many) spans were captured. With the Jaeger Tracer, you can enable the LoggingReporter, so that the spans are written also to the stdout. This way, if anything fails, you have some pointers on what may have gone wrong. For instance, if you don't see anything in the logs, then it's a problem with the instrumentation. If you see entries in the logs, but no spans in the collector/agent, then it's a networking issue.

**Comment by Beanow** [comment]
So fairly good news. I did some initial tests over at https://github.com/Beanow/jaegertracing-example-android

OpenTracing works out of the box, and I was able to report to a collector with jaeger-client.
I've put these notes in the readme as well, but these were some points that stood out.


- [The docs](https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/README.md)
recommend you use `Tracer tracer = Configuration.fromEnv().getTracer();`
which makes little sense for Android as we don't control the ENV vars.

- Duplicate META-INF file errors start happening once you include jaegers dependencies.
[Fixed with an exclude like this](https://stackoverflow.com/questions/44342455/more-than-one-file-was-found-with-os-independent-path-meta-inf-license#47509465)

- `<uses-permission android:name="android.permission.INTERNET"/>` required of course.

- Initializing the tracer on the main thread causes a `android.os.NetworkOnMainThreadException` due to `InetAddress.getLocalHost()` being called for the `ip` and `hostname` tags.
Either set those manually or initialize off the main thread.

- Randomness quality of `io.jaegertracing.internal.utils.Utils.uniqueId` is extremely low. Causing spanId collisions.

- Logging support requires an additional package `slf4j-android`
such as `org.slf4j:slf4j-android:1.7.25` from jcenter.

### Randomness quality

Here's an example of me setting a breakpoint to see the Long produced by
`Java6CompatibleThreadLocalRandom.current().nextLong();`

It almost seems tied to uptime as the later ones are less likely to collide.

```
-9201949609202618220 x3
-8619592029950413405 x3
-2026341794611032000 x2
-6396153114935662112 x3
 3161358585497756669 -
-2332688683697357959 -

== restart application ==

-9201949609202618220 x3
-6396153114935662112 x3
-8619592029950413405 x3
-6662863254555043445 x2
 8227148315116627998 -
-3110453752544509859 -
-2446405553574246131 -

== restart application ==

-9201949609202618220 x3
-8619592029950413405 x3
-6662863254555043445 x2
-6396153114935662112 x3
-2026341794611032000 x2
 1338145981130179361 -
 5205072654012701703 -
```

This ends up combining spans that weren't actually related.

![Collisions](https://raw.githubusercontent.com/Beanow/jaegertracing-example-android/master/img/span-id-collisions.png)

**Comment by yurishkuro** [comment]
I don't recall how we seed the RNG, but since we're using one instance per thread I wonder if we're using the same seed and getting the same sequences.

**Comment by yurishkuro** [comment]
Yeah, we're not using any seed at all, should use current time plus either threadname hash or some atomic counter.

https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/src/main/java/io/jaegertracing/internal/utils/Java6CompatibleThreadLocalRandom.java

**Comment by Beanow** [comment]
Seeding ThreadLocalRandom is actually not supported. Though the fallback does support it.
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadLocalRandom.html

> a ThreadLocalRandom is initialized with an internally generated seed that may not otherwise be modified.

While there's a possibility this is related to how kotlin implements coroutines, that would only explain collisions (i.e. race conditions) between those coroutines, not after restarting the application.

**Comment by jpkrohling** [comment]
I'm removing the `outreachy` label, as I think the research is done.

**Comment by Moes81** [comment]
So, is there any way to prevent the id collision?

**Comment by jpkrohling** [comment]
@Moes81 if you can come up with a PR that improves it, chances are very high that we can merge it.
